services/src/contexts/mod.rs

```diff
@@ -6,24 +6,39 @@ use crate::{
 use async_trait::async_trait;
 use std::sync::Arc;
 use tokio::sync::{RwLock, RwLockReadGuard, RwLockWriteGuard};
-
 mod in_memory;
 #[cfg(feature = "postgres")]
 mod postgres;
+use crate::datasets::storage::DataSetDB;
 
+use crate::users::user::UserId;
+use crate::util::config::{get_config_element, GdalSource};
+use geoengine_datatypes::dataset::DataSetId;
+use geoengine_operators::concurrency::ThreadPool;
+use geoengine_operators::engine::{
+    ExecutionContext, LoadingInfo, LoadingInfoProvider, QueryContext, VectorResultDescriptor,
+};
+use geoengine_operators::mock::MockDataSetDataSourceLoadingInfo;
+use geoengine_operators::source::OgrSourceDataset;
 pub use in_memory::InMemoryContext;
 #[cfg(feature = "postgres")]
 pub use postgres::PostgresContext;
+use std::borrow::Borrow;
+use std::path::PathBuf;
 
 type DB<T> = Arc<RwLock<T>>;
 
 /// A context bundles access to shared resources like databases and session specific information
 /// about the user to pass to the services handlers.
+// TODO: avoid locking the individual DBs here IF they are already thread safe (e.g. guaranteed by postgres)
```

Comment:
Yeah, this doesn't seem to be good to block here. Maybe you just need to allow operations and not exclusive access to the DB.

Reply:
I created a task https://trello.com/c/jCS0zj0c/193-optimize-locking-of-databases-in-context